### PR TITLE
Unions improvement: remove `as const`, add primitives support

### PR DESCRIPTION
**Pull request by kossnocorp**
This PR is based on https://github.com/kossnocorp/ofcoerce/pull/4 by @datner

It adds support for primitives in unions and also removes the need to use `as const` with literals.

### CODE REVIEW

**Comment by datner** [comment]
Responding to https://github.com/kossnocorp/ofcoerce/pull/4#issuecomment-2170970718 ,
The core problem is that unions can't be turned into arrays, just a typescript impossibility. So what I made was functionally a hack.. Theres also the issue that for the defined coerse you must provide the identity shape, but we want to accept the extendable shape. Meaning that this issue happens if you use arrays to track the type
```ts
type Foo = "a" | "b"
type Bar = {
  foo: Foo
}

coerse<Bar>($ => ({
  foo: $.Union("a","b") // Union<["a", "b"]> is not eq Union<["a" | "b"]>
})
```
but not using an array gimps you on the inference side for the same reason (in reverse though)!
You can either make two different types, `Union<"a" | "b">` for `Defined` and `TrackedUnion<["a", "b"]>` for `Inferred`
Or test for their unioned equality
The only sane way that could be done (from what I know) is to instantiate the type and then do the coerse because of how generics work
```ts
type UnionEq<A extends ReadonlyArray<any>, B extends ReadonlyArray<any>> =
  A[number] extends B[number] ? B[number] extends A[number] ? true : false : false

type Matching<Expected, Current> = ....

declare const coerse: <Self>() => 
  <Shape extends Self>
  (schema: true extends Matching<Self, Shape> ? Shape : never)
    => Core.Coerser<Shape>

coerse<Bar>()($ => ({
  foo: $.Union("a", "b")
})
```

basically Unions are a PITA and might need a rethink on the type-side
